crates/turborepo-lib/src/shim.rs

```diff
@@ -153,24 +179,82 @@ impl ShimArgs {
             }
         }
 
-        if found_cwd_flag {
-            Err(Error::EmptyCwd(Backtrace::capture()))
-        } else {
-            let invocation_dir = AbsoluteSystemPathBuf::cwd()?;
-            let cwd = cwd.unwrap_or_else(|| invocation_dir.clone());
-
-            Ok(ShimArgs {
-                cwd,
-                invocation_dir,
-                skip_infer,
-                verbosity,
-                force_update_check,
-                remaining_turbo_args,
-                forwarded_args,
-                color,
-                no_color,
-            })
+        if let Some(idx) = cwd_flag_idx {
+            let start_idx = Self::get_indices_in_args_string(vec![idx], env::args().skip(1));
+            let args_string = env::args().skip(1).join(" ");
+
+            return Err(Error::EmptyCwd {
+                backtrace: Backtrace::capture(),
+                args_string,
+                flag_range: (start_idx[0], "--cwd".len()).into(),
+            });
+        }
+
+        if cwds.len() > 1 {
+            let indices = Self::get_indices_in_args_string(
+                cwds.iter().map(|(_, idx)| *idx).collect(),
+                env::args().skip(1),
+            );
+            let args_string = env::args().skip(1).join(" ");
+
+            return Err(Error::MultipleCwd(Box::new(MultipleCwd {
+                backtrace: Backtrace::capture(),
+                args_string,
+                flag1: indices.first().map(|i| (*i, "--cwd".len()).into()),
+                flag2: indices.get(1).map(|i| (*i, "--cwd".len()).into()),
+                flag3: indices.get(2).map(|i| (*i, "--cwd".len()).into()),
+                flag4: indices.get(3).map(|i| (*i, "--cwd".len()).into()),
+            })));
         }
+
+        let invocation_dir = AbsoluteSystemPathBuf::cwd()?;
+        let cwd = cwds
+            .pop()
+            .map(|(cwd, _)| cwd)
+            .unwrap_or_else(|| invocation_dir.clone());
+
+        Ok(ShimArgs {
+            cwd,
+            invocation_dir,
+            skip_infer,
+            verbosity,
+            force_update_check,
+            remaining_turbo_args,
+            forwarded_args,
+            color,
+            no_color,
+        })
+    }
+
+    /// Takes a list of indices into a Vec of arguments, i.e. ["--graph", "foo",
+    /// "--cwd"] and converts them into indices into the string of those
+    /// arguments, i.e. "-- graph foo --cwd"
+    pub(crate) fn get_indices_in_args_string(
+        mut args_indices: Vec<usize>,
+        args: impl Iterator<Item = impl Into<String>>,
+    ) -> Vec<usize> {
+        // Sort the indices to keep the invariant
+        // that if i > j then output[i] > output[j]
+        args_indices.sort();
```

Comment:
Do we need to require callers to keep up this invariant? e.g. If I pass in `vec![2, 1]` and get out `vec![3, 6]` would I be in a bad state when I try to map the indices to the args.

Reply:
Well I think the idea is you shouldn't be mapping these to args, since they're now indices into the string. And unless you parse the string it's not really easy to map the indices to the args.